source/world_builder/features/oceanic_plate_models/composition/water_content.cc

```diff
@@ -0,0 +1,239 @@
+/*
+  Copyright (C) 2018 - 2024 by the authors of the World Builder code.
+
+  This file is part of the World Builder.
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU Lesser General Public License as published
+   by the Free Software Foundation, either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "world_builder/features/oceanic_plate_models/composition/water_content.h"
+
+#include "world_builder/kd_tree.h"
+#include "world_builder/nan.h"
+#include "world_builder/types/array.h"
+#include "world_builder/types/double.h"
+#include "world_builder/types/object.h"
+#include "world_builder/types/one_of.h"
+#include "world_builder/types/unsigned_int.h"
+#include "world_builder/types/value_at_points.h"
+#include "world_builder/utilities.h"
+#include "world_builder/world.h"
+
+
+namespace WorldBuilder
+{
+
+  using namespace Utilities;
+
+  namespace Features
+  {
+    namespace OceanicPlateModels
+    {
+      namespace Composition
+      {
+        WaterContent::WaterContent(WorldBuilder::World *world_)
+          :
+          min_depth(NaN::DSNAN),
+          max_depth(NaN::DSNAN),
+          density(NaN::DSNAN)
+        {
+          this->world = world_;
+          this->name = "water content";
+        }
+
+        WaterContent::~WaterContent()
+          = default;
+
+        void
+        WaterContent::declare_entries(Parameters &prm, const std::string & /*unused*/)
+        {
+          // Document plugin and require entries if needed.
+          // Add compositions to the required parameters.
+          prm.declare_entry("", Types::Object({"compositions"}),
+                            "WaterContent compositional model. Sets bound water content as a compositional field.");
```

Comment:
Could you extend this, explaining it is the Tian 2019 model, and add the changes you made to make it work? Also in the one below. Basically what you have written in the class description.